Prove that Scott locales of Scott domains are spectral

ayberkt opened the pull request:
In PR #201, the Scott locale of a large and locally small dcpo was shown to be locally small.

This PR builds on PR #201, and adds a further proof that the large and locally small Scott locale of a large and locally small Scott domain is spectral, under the additional assumption that the existence of binary upper bounds is decidable for compact elements in the domain in consideration.

ayberkt commented:
@martinescardo The proof of spectrality of the Scott locale of a large and locally small Scott domain is now complete!

/cc @tomdjong

tomdjong commented:
@ayberkt If you would like me to review, just let me know when it's ready. (I still noticed a number of `TODO`s upon skimming the changes.)

martinescardo commented on source/DomainTheory/BasesAndContinuity/ScottDomain.lagda:
```diff
@@ -0,0 +1,114 @@
+start-date:  2023-10-26
+---
+
+Ayberk Tosun.
```
Do you need lines 7-9 if you have lines 1-5? By the way, if we adopt your header-suggestion, I would suggest to add a new field `dates-modified:` as well. In any case, we need to discuss this better on another occasion.

ayberkt replied:
> Do you need lines 7-9 if you have lines 1-5?

I also included lines 7-9 because the preamble in lines 1-5 is not a convention that we agreed on.

The idea was that, even if we end up not adopting this kind of a convention, it doesn’t hurt to put in a bit of additional information that can be ignored by the reader. However, I can remove it if you don’t like it.

I plan to implement a prototype of formatted HTML generation on my own development, and I need this information to be available in a standard format so that I can e.g. display the title as the title of the HTML page.

martinescardo commented on source/DomainTheory/BasesAndContinuity/ScottDomain.lagda:
```diff
@@ -0,0 +1,114 @@
+open import UF.PropTrunc
+open import UF.SubtypeClassifier
+open import UF.Subsingletons
+-- open import UF.Subsingletons-FunExt
```
Why are these two lines commented out?

ayberkt replied:
Just forgotten lines. Fixed.

martinescardo commented on source/DomainTheory/Topology/ScottTopologyProperties.lagda:
```diff
@@ -0,0 +1,225 @@
+
+\begin{code}
+
+ characterization-of-scott-opens : (U : 𝓟 {𝓥} ⟨ 𝓓 ⟩)
```
Maybe write this as follows for the sake of legibility:
```
characterization-of-scott-opens 
 : (U : 𝓟 {𝓥} ⟨ 𝓓 ⟩)
→ (is-scott-open U ⇒ (Ɐ x ꞉ ⟨ 𝓓 ⟩ , U x ⇔ join-of-compact-opens U x)) holds
```

ayberkt replied:
Good idea. Done now.

martinescardo commented on source/DomainTheory/index.lagda:
```diff
@@ -137,10 +139,12 @@ import DomainTheory.ScottModelOfPCF.ScottModelOfPCF -- (2)
 {- Topology (by Ayberk Tosun)
```
@tomdjong Please check how to best add authorship in this file.

martinescardo commented on source/Locales/Frame.lagda:
```diff
@@ -549,6 +549,21 @@ syntax binary-join F x y = x ∨[ F ] y
 
+\begin{code}
+
+ub₂-implies-ub-for-binary-family : {𝓤 𝓥 𝓦 : Universe} {A : 𝓤  ̇}
```
Indent as suggested above in this review.

martinescardo commented on source/Locales/ScottLocale/Properties.lagda:
```diff
@@ -0,0 +1,86 @@
+ hiding   (is-directed)
+open import DomainTheory.Basics.Pointed                      pt fe 𝓤
+ renaming (⊥ to ⊥d)
+open import DomainTheory.Topology.ScottTopology        pt fe 𝓤
```
(Why are the `pt`s here not indented as the previous ones?)

martinescardo commented on source/Locales/ScottLocale/Properties.lagda:
```diff
@@ -0,0 +1,86 @@
+\begin{code}
+
+module ScottLocaleProperties (𝓓 : DCPO {𝓤 ⁺} {𝓤})
+                             (hl : has-least (underlying-order 𝓓))
```
Probably break line and indent one space after the module name.

ayberkt replied:
Done.

martinescardo commented on source/Locales/ScottLocale/Properties.lagda:
```diff
@@ -0,0 +1,86 @@
+
+ open DefnOfScottLocale 𝓓 𝓤 pe using (_⊆ₛ_)
+
+ ↑⊥-is-below-𝟏 : (𝟏[ 𝒪 Σ⦅𝓓⦆ ] ⊆ₛ ↑ˢ[ ⊥ᴰ , ⊥κ ]) holds
```
I think it is worth adding a comment before line 77 explaining what line 77 means. I've found this not very readable.

ayberkt replied:
Done.

martinescardo commented on source/Locales/ScottLocale/Properties.lagda:
```diff
@@ -0,0 +1,86 @@
+ ↑⊥-is-below-𝟏 : (𝟏[ 𝒪 Σ⦅𝓓⦆ ] ⊆ₛ ↑ˢ[ ⊥ᴰ , ⊥κ ]) holds
+ ↑⊥-is-below-𝟏 = bottom-principal-filter-is-top 𝟏[ 𝒪 Σ⦅𝓓⦆ ]
+
+ ↑⊥-is-top : ↑ˢ[ ⊥ᴰ , ⊥κ ] ＝ 𝟏[ 𝒪 Σ⦅𝓓⦆ ]
```
Probably here too.

ayberkt replied:
Done.

martinescardo commented on source/Locales/ScottLocale/ScottLocalesOfScottDomains.lagda:
```diff
@@ -0,0 +1,822 @@
+---
+
+In this module, we prove that the Scott locale of any Scott domain is a spectral
+locale (provided that the domain in consideration is large and locally small and
```
The comment is brackets is very important and so please remove the brackets.

martinescardo commented on source/Locales/ScottLocale/ScottLocalesOfScottDomains.lagda:
```diff
@@ -0,0 +1,822 @@
+open import Locales.Spectrality.SpectralLocale               pt fe
+
+open import Locales.SmallBasis pt fe sr
+
```
Remove space.

martinescardo commented on source/Locales/ScottLocale/ScottLocalesOfScottDomains.lagda:
```diff
@@ -0,0 +1,822 @@
+open import Locales.SmallBasis pt fe sr
+
+open AllCombinators pt fe
+
```
Remove space.

martinescardo commented on source/Locales/ScottLocale/ScottLocalesOfScottDomains.lagda:
```diff
@@ -0,0 +1,822 @@
+
+decidability-condition : (𝓓 : DCPO {𝓤 ⁺} {𝓤}) → 𝓤 ⁺  ̇
+decidability-condition 𝓓 = (c d : ⟨ 𝓓 ⟩∙) →
+                            is-compact 𝓓 c → is-compact 𝓓 d →
```
One implication only in each line, when there are multiple lines, please.

martinescardo commented on source/Locales/ScottLocale/ScottLocalesOfScottDomains.lagda:
```diff
@@ -0,0 +1,822 @@
+  1. The dcpo `𝓓` in consideration is large and locally small.
+  2. It is pointed.
+  3. It has a specified small compact basis.
+  4. It satisfies the aforementioned decidability condition.
```
Please repeat the condition to make the file easier to read.

ayberkt replied:
> Do we really need a specified basis?

No we don’t. This is what I do in the second module `SpectralScottLocaleConstruction₂`: I use the split support result to construct a basis out of the unspecified basis, and then I use this module to obtain the proof.

I should probably use a name more descriptive than `SpectralScottLocaleConstruction₂`. I’ll maybe call it `SpectralScottLocaleConstructionFromUnspecifiedBasis`.

martinescardo replied:
The you should add a comment to that extent.

martinescardo commented on source/Locales/ScottLocale/ScottLocalesOfScottDomains.lagda:
```diff
@@ -0,0 +1,822 @@
+
+\end{code}
+
+We denote by `Σ𝓓` the large and locally small Scott locale of the dcpo `𝓓`.
```
I know that the usual notation for the Scott topology of `𝓓` is `Σ𝓓` (and I told you so verbally). However, I think it is very confusing to use `Σ` for this in the context of dependent type theory. So please rename. But also notice that the notation you use in the comments doesn't match the notation you use in the Agda code.

martinescardo commented on source/Locales/ScottLocale/ScottLocalesOfScottDomains.lagda:
```diff
@@ -0,0 +1,822 @@
+
+ open is-small-compact-basis scb
+
+
```
Too many blank lines.

ayberkt replied:
Fixed.

martinescardo commented on source/Locales/ScottLocale/ScottLocalesOfScottDomains.lagda:
```diff
@@ -0,0 +1,822 @@
+
+ open DefnOfScottTopology 𝓓 𝓤
+ open Properties 𝓓
+
```
I don't think you need blank lines between `open` statements.

ayberkt replied:
Fixed.

martinescardo reviewed (changes requested):
Thanks, @ayberkt . I made some minor suggestions. I've also asked @tomdjong to review this, as it is based on his domain-theory development.

ayberkt commented:
@tomdjong Yes I think it'll be a good idea for me to address some TODOs before you take a look. I'll address @martinescardo's comments first and then send you a review request.

ayberkt commented on source/Locales/ScottLocale/ScottLocalesOfScottDomains.lagda:
```diff
@@ -0,0 +1,822 @@
+ open is-small-compact-basis scb
+
+
+ ϟ : (b : B) → is-compact 𝓓 (β b)
```
Ah sorry, I got into a bad habit of using [the lost letters of the Greek alphabet](https://steve-lovelace.com/greek-alphabet/) when I run out of Greek letters.

ayberkt replied:
I’ll fix this.

ayberkt commented on source/Locales/ScottLocale/ScottLocalesOfScottDomains.lagda:
```diff
@@ -0,0 +1,822 @@
+    † (inr r) = ∣ inr (𝜸₀-is-upwards-closed bs x y r q) ∣
+
+ 𝜸₀-is-inaccessible-by-directed-joins : (ks : List B)
+                                             → is-inaccessible-by-directed-joins
```
Fixed.

ayberkt commented:
@martinescardo @tomdjong

Status update: I’ve been a bit late in addressing the comments and then sending a review request to @tomdjong as  I promised. This is because I decided to include a proof of spectrality the Sierpiński locale in this PR. I think this will be nice to include as a small example of how to use the main proof added by the PR.

ayberkt commented:
@martinescardo @tomdjong: This PR is now ready to review.

In addition to the contributions previously summarized, I have also added a proof that the Sierpiński dcpo is a Scott domain in which the upper-boundedness of compact elements is decidable. These are both trivial because the Sierpiński dcpo is small complete, and its type of compact elements has all binary joins. Using this, I defined the patch of Sierpiński to showcase the main proof added in this PR: the Scott locale of any Scott domain, satisfying the aforementioned decidability condition, is a spectral locale.

tomdjong commented:
@ayberkt Might it possible to split this PR into two: a first one on domain theory and a second one on locale theory? This would make it easier to review.

ayberkt commented:
> @ayberkt Might it possible to split this PR into two: a first one on domain theory and a second one on locale theory? This would make it easier to review.

Sure I’d be happy to do that.

tomdjong commented:
@ayberkt Since we're splitting this up, are you with me making this a draft?

ayberkt commented:
> @ayberkt Since we're splitting this up, are you with me making this a draft?

Yes good idea. In fact, I’ll just close it.

ayberkt commented:
This PR has been broken up into three PRs.

1. PR #240 containing the additions to `DomainTheory` development.
2. PR #255 containing the actual proof of spectrality of the Scott locale of a Scott domain.
3. PR #256 containing the proof that the Sierpiński locale is spectral, as an example of the proof that the Scott locale of any Scott domain (satisfying a certain decidability condition) is spectral.
